sway-ir/src/optimize/misc_demotion.rs

```diff
@@ -445,39 +451,245 @@ fn wide_binary_op_demotion(context: &mut Context, function: Function) -> Result<
+
+        let block_instrs = &mut context.blocks[block.0].instructions;
+
+        block_instrs.insert(idx, wide_op);
+        block_instrs.insert(idx, get_result_local);
+
+        if arg2_needs_insert {
+            block_instrs.insert(idx, get_arg2);
+        }
+
+        if arg1_needs_insert {
+            block_instrs.insert(idx, get_arg1);
+        }
+
+        //rhs
+        if let Some((get_rhs_local, store_rhs_local)) = rhs_store {
+            block_instrs.insert(idx, store_rhs_local);
+            block_instrs.insert(idx, get_rhs_local);
+        }
+
+        // Only for MOD
+        if let Some(get_local_zero) = get_local_zero {
+            block_instrs.insert(idx, get_local_zero);
+        }
+
+        // lhs
+        if let Some((get_lhs_local, store_lhs_local)) = lhs_store {
+            block_instrs.insert(idx, store_lhs_local);
+            block_instrs.insert(idx, get_lhs_local);
+        }
+    }
+
+    Ok(true)
+}
+
+/// Find all cmp operations on types bigger than 64 bits
+/// and demote them to fuel specific `wide cmp ops`, that
+/// work only on pointers
+fn wide_cmp_demotion(context: &mut Context, function: Function) -> Result<bool, IrError> {
+    // Find all cmp on wide operators
+    let candidates = function
+        .instruction_iter(context)
+        .filter_map(|(block, instr_val)| {
+            let instr = instr_val.get_instruction(context)?;
+
+            if let Instruction::Cmp(op, arg1, arg2) = instr {
+                let arg1_type = arg1
+                    .get_type(context)
+                    .and_then(|x| x.get_uint_width(context));
+                let arg2_type = arg2
+                    .get_type(context)
+                    .and_then(|x| x.get_uint_width(context));
+
+                match (arg1_type, arg2_type) {
+                    (Some(256), Some(256)) => {
+                        use Predicate::*;
+                        match op {
+                            Equal => Some((block, instr_val)),
+                            _ => todo!(),
+                        }
+                    }
+                    _ => None,
+                }
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<_>>();
+
+    if candidates.is_empty() {
+        return Ok(false);
+    }
+
+    // Get ptr to each arg
+    for (block, cmp_instr_val) in candidates {
+        let Instruction::Cmp (op, arg1, arg2) = cmp_instr_val
+            .get_instruction(context)
+            .cloned()
+            .unwrap() else {
+                continue;
+            };
+
+        let cmp_op_metadata = cmp_instr_val.get_metadata(context);
+
+        let arg1_ty = arg1.get_type(context).unwrap();
+        let arg1_metadata = arg1.get_metadata(context);
+        let arg2_ty = arg2.get_type(context).unwrap();
+        let arg2_metadata = arg2.get_metadata(context);
+
+        // If arg1 is not a pointer, store it to a local
+        let lhs_store = if !arg1_ty.is_ptr(context) {
```

Comment:
You can use [`then`](https://doc.rust-lang.org/std/primitive.bool.html#method.then) here

Reply:
Done.